chalk-parse/src/ast.rs

```diff
@@ -1,70 +1,140 @@
 use lalrpop_intern::InternedString;
 use std::fmt;
 
+/// A span is a range of characters in a string where an error occured.
 #[derive(Copy, Clone, Debug, PartialEq, Eq)]
 pub struct Span {
     pub lo: usize,
     pub hi: usize,
 }
 
 impl Span {
+    /// Make a new span.
     pub fn new(lo: usize, hi: usize) -> Self {
         Span { lo: lo, hi: hi }
     }
 }
 
+/// The top-level container for a parsed .chalk file.
 pub struct Program {
     pub items: Vec<Item>
 }
 
+/// The `Item` enum represents a statement in a .chalk file.
 pub enum Item {
     StructDefn(StructDefn),
     TraitDefn(TraitDefn),
     Impl(Impl),
     Clause(Clause),
 }
 
+/// Represents a rust struct.
 pub struct StructDefn {
+    /// The name of the structure.
     pub name: Identifier,
+
+    /// The type and lifetime parameters for this struct. They are stored in the
+    /// order they appear in the struct definition.
     pub parameter_kinds: Vec<ParameterKind>,
+
+    /// Any bounds on the type parameters are stored here.
     pub where_clauses: Vec<QuantifiedWhereClause>,
+
+    /// The field names and types for this struct.
     pub fields: Vec<Field>,
+
+    /// Any modifiers on this struct.
     pub flags: StructFlags,
 }
 
+/// This structure contains any modifiers attached to a rust struct. Currently,
+/// only the `extern` modifier is tracked.
 pub struct StructFlags {
+    /// This flag is true if the struct is `extern`.
     pub external: bool,
 }
 
+/// Represents a rust trait.
 pub struct TraitDefn {
+    /// The name of the trait.
     pub name: Identifier,
+
+    /// The type and lifetime parameters for this trait. They are stored in the
+    /// order they appear in the trait definition.
     pub parameter_kinds: Vec<ParameterKind>,
+
+    /// Any bounds on the type parameters are stored here.
     pub where_clauses: Vec<QuantifiedWhereClause>,
+
+    /// Associated types for this trait.
     pub assoc_ty_defns: Vec<AssocTyDefn>,
+
+    /// Any modifiers on this trait.
     pub flags: TraitFlags,
 }
 
+/// This structure contains any flags attached to a rust trait.
 pub struct TraitFlags {
+
+    /// This is true if the trait is `#[auto]`.
     pub auto: bool,
+
+    /// This is true if the trait is `#[marker]`.
     pub marker: bool,
+
+    /// This is true if the trait is `extern`.
```

Comment:
"That means it's defined outside the 'current crate', which can be significant to coherence. See <https://rust-lang.github.io/rfcs/1023-rebalancing-coherence.html>."